File: drce_loc.cpp
#include "drce_loc.hpp"

// ----------------------------------------------------------------------------
// Step 1: Image segmentation + Block_mean / Block_std
// ----------------------------------------------------------------------------
void DRCELOC::computeBlockStats(const cv::Mat& img32f, cv::Mat& blockMean, cv::Mat& blockStd) {
    const int M = img32f.rows, N = img32f.cols;
    const int X = params_.blockRows, Y = params_.blockCols;

    blockMean.create(X, Y, CV_64F);
    blockStd.create(X, Y, CV_64F);

    for (int bx = 0; bx < X; ++bx) {
        int r0 = static_cast<int>(std::round(bx * (double)M / X));
        int r1 = static_cast<int>(std::round((bx + 1) * (double)M / X));
        r1 = std::max(r1, r0 + 1);
        for (int by = 0; by < Y; ++by) {
            int c0 = static_cast<int>(std::round(by * (double)N / Y));
            int c1 = static_cast<int>(std::round((by + 1) * (double)N / Y));
            c1 = std::max(c1, c0 + 1);

            cv::Rect roi(c0, r0, c1 - c0, r1 - r0);
            cv::Scalar mean, stddev;
            cv::meanStdDev(img32f(roi), mean, stddev);

            blockMean.at<double>(bx, by) = mean[0];
            blockStd.at<double>(bx, by)  = stddev[0];
        }
    }
}

// ----------------------------------------------------------------------------
// Step 3: Gaussian upsampling блочной карты (X x Y) до полного разрешения
// (M x N), Eq. 2 / Eq. 3.
//
// Точная формула:
//   N(i,j) = sum_{x,y} V(x,y)*exp(-d((i,j),(x,y))^2/delta2) / sum_{x,y} exp(-d^2/delta2)
//
// Разделимость (см. комментарий в drce_loc.hpp и Eq. 15 статьи):
//   exp(-d^2/delta2) = exp(-dRow(i,x)^2/delta2) * exp(-dCol(j,y)^2/delta2)
//
// Тогда числитель = Wrow (MxX) * V (XxY) * Wcol^T (YxN)  -> M x N,
//      знаменатель = rowSum(Wrow) (Mx1) * colSum(Wcol)^T (1xN) -> M x N (outer product),
// что считается двумя cv::gemm вместо O(M*N*X*Y) перебора.
// ----------------------------------------------------------------------------
cv::Mat DRCELOC::gaussianUpsample(const cv::Mat& blockMap64f, int M, int N, double delta2) {
    const int X = blockMap64f.rows, Y = blockMap64f.cols;

    // Центры блоков в координатах полного изображения.
    std::vector<double> rowCenters(X), colCenters(Y);
    for (int bx = 0; bx < X; ++bx) rowCenters[bx] = (bx + 0.5) * (double)M / X;
    for (int by = 0; by < Y; ++by) colCenters[by] = (by + 0.5) * (double)N / Y;

    cv::Mat Wrow(M, X, CV_64F);   // Wrow(i,x) = exp(-(i-cx)^2/delta2)
    cv::Mat Wcol(N, Y, CV_64F);   // Wcol(j,y) = exp(-(j-cy)^2/delta2)

    for (int i = 0; i < M; ++i)
        for (int bx = 0; bx < X; ++bx) {
            double d = i - rowCenters[bx];
            Wrow.at<double>(i, bx) = std::exp(-(d * d) / delta2);
        }
    for (int j = 0; j < N; ++j)
        for (int by = 0; by < Y; ++by) {
            double d = j - colCenters[by];
            Wcol.at<double>(j, by) = std::exp(-(d * d) / delta2);
        }

    // Числитель: (Wrow * V) * Wcol^T
    cv::Mat temp = Wrow * blockMap64f;      // M x Y
    cv::Mat numerator = temp * Wcol.t();    // M x N

    // Знаменатель: rowSum outer colSum
    cv::Mat rowSum, colSum;
    cv::reduce(Wrow, rowSum, 1, cv::REDUCE_SUM); // M x 1
    cv::reduce(Wcol, colSum, 1, cv::REDUCE_SUM); // N x 1
    cv::Mat denom = rowSum * colSum.t();         // M x N (outer product)

    cv::Mat result;
    cv::divide(numerator, denom, result);
    result.convertTo(result, CV_32F);
    return result;
}

// ----------------------------------------------------------------------------
// Step 4: Self-guided filter (p = I, guide = I), окно w x w, Eq. 4.
// Стандартная реализация через box filter (He et al., 2012), специализация
// для self-guided случая: a_k = var_I/(var_I+eps); b_k = mean_I*(1-a_k).
// ----------------------------------------------------------------------------
void DRCELOC::selfGuidedFilter(const cv::Mat& img32f, cv::Mat& base, cv::Mat& detail, cv::Mat& aCoeff) {
    const int r = params_.guidedRadius;
    const cv::Size ksize(2 * r + 1, 2 * r + 1);
    const double eps = params_.guidedEps;

    cv::Mat meanI, meanII, varI;
    cv::boxFilter(img32f, meanI, CV_32F, ksize);
    cv::boxFilter(img32f.mul(img32f), meanII, CV_32F, ksize);
    varI = meanII - meanI.mul(meanI);

    cv::Mat a = varI / (varI + eps);
    cv::Mat b = meanI - a.mul(meanI); // = meanI*(1-a)

    cv::Mat meanA, meanB;
    cv::boxFilter(a, meanA, CV_32F, ksize);
    cv::boxFilter(b, meanB, CV_32F, ksize);

    base = meanA.mul(img32f) + meanB;
    detail = img32f - base;
    aCoeff = a;
}

// ----------------------------------------------------------------------------
// Step 5: Global brightness guide map, Eq. 5.
// ----------------------------------------------------------------------------
cv::Mat DRCELOC::computeBrightnessGuide(const cv::Mat& img32f) {
    cv::Scalar mean, stddev;
    cv::meanStdDev(img32f, mean, stddev);

    cv::Mat Igc = (255.0 * (img32f - mean[0])) / (stddev[0] + params_.lambda) + params_.bright;
    return Igc;
}

// ----------------------------------------------------------------------------
// Основной конвейер: Steps 1-8.
// ----------------------------------------------------------------------------
cv::Mat DRCELOC::process(const cv::Mat& src) {
    CV_Assert(!src.empty());
    CV_Assert(src.channels() == 1);

    cv::Mat img32f;
    src.convertTo(img32f, CV_32F);

    if (params_.normalizeInputTo255) {
        double lo, hi;
        cv::minMaxLoc(img32f, &lo, &hi);
        if (hi - lo > 1e-6) {
            img32f = (img32f - lo) * (255.0 / (hi - lo));
        }
    }

    const int M = img32f.rows, N = img32f.cols;
    const int X = params_.blockRows, Y = params_.blockCols;

    // --- Step 1-2 ---
    computeBlockStats(img32f, dbg_.blockMean, dbg_.blockStd);
    cv::Mat blockStretch = 255.0 / (dbg_.blockStd + params_.bias); // Eq. 1
    dbg_.blockStretch = blockStretch;

    // --- Step 3 ---
    double delta2 = params_.delta2;
    if (delta2 <= 0.0) {
        double blockArea = ((double)M / X) * ((double)N / Y);
        delta2 = 1.5 * blockArea; // статья: delta^2 >= размер блока (Section 2.4)
    }
    dbg_.NStretch = gaussianUpsample(blockStretch, M, N, delta2);
    dbg_.NMean    = gaussianUpsample(dbg_.blockMean, M, N, delta2);

    // --- Step 4 ---
    selfGuidedFilter(img32f, dbg_.gfBase, dbg_.gfDetail, dbg_.gfA);

    // --- Step 5 ---
    dbg_.Igc = computeBrightnessGuide(img32f);

    // --- Step 6: Eq. 6 ---
    // Base_out = k1*NStretch*(Iin - NMean) + k2*Igc
    cv::Mat baseOut = params_.k1 * dbg_.NStretch.mul(img32f - dbg_.NMean) + params_.k2 * dbg_.Igc;
    dbg_.baseOut = baseOut;

    // --- Step 7: Eq. 7 (интерпретация, см. комментарий в drce_loc.hpp) ---
    cv::Mat absA = cv::abs(dbg_.gfA);
    double maxA;
    cv::minMaxLoc(absA, nullptr, &maxA);
    if (maxA < 1e-9) maxA = 1.0;
    cv::Mat normA = absA / maxA; // 0..1, мера локальной "детализированности"
    cv::Mat mask = params_.gLow + (params_.gHigh - params_.gLow) * normA;
    cv::Mat detailOut = dbg_.gfDetail.mul(absA).mul(mask);
    dbg_.detailOut = detailOut;

    // --- Step 8: Eq. 8 ---
    cv::Mat out32f = baseOut + params_.dde * detailOut;

    cv::Mat out8u;
    cv::Mat clipped;
    cv::min(cv::max(out32f, 0.0), 255.0, clipped);
    clipped.convertTo(out8u, CV_8U);
    return out8u;
}

File: python/drce_loc.py
# -*- coding: utf-8 -*-
"""
Порт DRCE-LOC на Python (numpy + OpenCV).

Реализация по статье:
  Zhu Y., Zhou Y., Jin W., Zhang L., Wu G., Shao Y.
  "A Low-Delay Dynamic Range Compression and Contrast Enhancement Algorithm
   Based on an Uncooled Infrared Sensor with Local Optimal Contrast"
  Sensors 2023, 23, 8860.

Соответствие шагам статьи (Section 2.1, Figure 2):
  Step 1  computeBlockStats()      -> Block_mean, Block_std
  Step 2  computeStretchCoeff()    -> Stretch_para  (Eq. 1)
  Step 3  gaussianUpsample()       -> NStretch, NMean (Eq. 2,3)
  Step 4  selfGuidedFilter()       -> Base(GF), Detail, a_coeff (Eq. 4)
  Step 5  computeBrightnessGuide() -> Igc           (Eq. 5)
  Step 6  process(): Base_out                              (Eq. 6)
  Step 7  process(): Detail_out (noise mask)              (Eq. 7)
  Step 8  process(): Iout = Base_out + DDE*Detail_out     (Eq. 8)

Вход: одноканальный массив numpy любой глубины (uint8/uint16/float32).
Выход: CV_8U массив того же размера (numpy uint8).
"""

import cv2
import numpy as np


class DRCELOC:
    class Params:
        def __init__(self):
            # Предварительная нормализация входа в номинальный диапазон 0..255
            # (min-max по кадру) до блочной статистики.
            self.normalizeInputTo255 = True

            # --- Шаг 1-3: блочная статистика и её апсемплинг ---
            self.blockRows = 12     # X: число блоков по вертикали
            self.blockCols = 16     # Y: число блоков по горизонтали
            self.bias = 200.0       # Bias, Eq. 1
            self.delta2 = -1.0      # delta^2, Eq. 2-3 (<=0 -> автоматически)

            # --- Шаг 4: guided filter (self-guided, окно 5x5 -> radius=2) ---
            self.guidedRadius = 2
            self.guidedEps = 100.0  # регуляризация в Eq. 4 (epsilon)

            # --- Шаг 5: глобальная яркостная направляющая карта ---
            self.lambda_ = 200.0    # lambda, Eq. 5
            self.bright = 128.0     # Bright, Eq. 5

            # --- Шаг 6: компрессия ДД + контрастное усиление ---
            self.k1 = 1.0           # вес локального контраста, Eq. 6
            self.k2 = 0.4           # вес глобального контраста / подавление гало, Eq. 6

            # --- Шаг 7: шумовая маска детального слоя ---
            self.gLow = 0.3         # минимальный gain в однородных зонах
            self.gHigh = 1.2        # максимальный gain в детализированных зонах

            # --- Шаг 8: финальный синтез ---
            self.dde = 1.0          # Detail enhancement factor, Eq. 8

    def __init__(self, params=None):
        self.params = params if params is not None else DRCELOC.Params()
        self.dbg = {}

    @staticmethod
    def _box_filter(img32f, ksize):
        # Прямый аналог cv::boxFilter(src, CV_32F, ksize).
        return cv2.boxFilter(img32f, -1, ksize)

    def compute_block_stats(self, img32f):
        M, N = img32f.shape[:2]
        X, Y = self.params.blockRows, self.params.blockCols
        block_mean = np.zeros((X, Y), dtype=np.float64)
        block_std = np.zeros((X, Y), dtype=np.float64)
        for bx in range(X):
            r0 = int(round(bx * M / X))
            r1 = int(round((bx + 1) * M / X))
            r1 = max(r1, r0 + 1)
            for by in range(Y):
                c0 = int(round(by * N / Y))
                c1 = int(round((by + 1) * N / Y))
                c1 = max(c1, c0 + 1)
                roi = img32f[r0:r1, c0:c1]
                block_mean[bx, by] = roi.mean()
                block_std[bx, by] = roi.std()  # population std (ddof=0), как meanStdDev
        return block_mean, block_std

    def gaussian_upsample(self, block_map, M, N, delta2):
        X, Y = block_map.shape[:2]
        row_centers = np.array([(bx + 0.5) * M / X for bx in range(X)])
        col_centers = np.array([(by + 0.5) * N / Y for by in range(Y)])

        i = np.arange(M)[:, None]          # M x 1
        j = np.arange(N)[:, None]          # N x 1
        wrow = np.exp(-((i - row_centers) ** 2) / delta2)   # M x X
        wcol = np.exp(-((j - col_centers) ** 2) / delta2)   # N x Y

        numerator = (wrow @ block_map) @ wcol.T              # M x N
        row_sum = wrow.sum(axis=1)[:, None]                  # M x 1
        col_sum = wcol.sum(axis=1)[None, :]                  # 1 x N
        denom = row_sum * col_sum                            # M x N (outer product)

        result = numerator / denom
        return result.astype(np.float32)

    def self_guided_filter(self, img32f):
        r = self.params.guidedRadius
        ksize = (2 * r + 1, 2 * r + 1)
        eps = self.params.guidedEps

        mean_i = self._box_filter(img32f, ksize)
        mean_ii = self._box_filter(img32f * img32f, ksize)
        var_i = mean_ii - mean_i * mean_i

        a = var_i / (var_i + eps)
        b = mean_i - a * mean_i

        base = self._box_filter(a, ksize) * img32f + self._box_filter(b, ksize)
        detail = img32f - base
        return base, detail, a

    def compute_brightness_guide(self, img32f):
        mean = float(img32f.mean())
        std = float(img32f.std())
        return (255.0 * (img32f - mean)) / (std + self.params.lambda_) + self.params.bright

    def process(self, src):
        assert src.ndim == 2 and src.size > 0, "DRCELOC: ожидается одноканальное непустое изображение"
        img = src.astype(np.float32)

        if self.params.normalizeInputTo255:
            lo, hi = float(img.min()), float(img.max())
            if hi - lo > 1e-6:
                img = (img - lo) * (255.0 / (hi - lo))

        M, N = img.shape[:2]
        X, Y = self.params.blockRows, self.params.blockCols

        block_mean, block_std = self.compute_block_stats(img)
        block_stretch = 255.0 / (block_std + self.params.bias)   # Eq. 1
        self.dbg['blockMean'] = block_mean
        self.dbg['blockStd'] = block_std
        self.dbg['blockStretch'] = block_stretch

        delta2 = self.params.delta2
        if delta2 <= 0.0:
            block_area = (M / X) * (N / Y)
            delta2 = 1.5 * block_area
        self.dbg['NStretch'] = self.gaussian_upsample(block_stretch, M, N, delta2)
        self.dbg['NMean'] = self.gaussian_upsample(block_mean, M, N, delta2)

        gf_base, gf_detail, gf_a = self.self_guided_filter(img)
        self.dbg['gfBase'] = gf_base
        self.dbg['gfDetail'] = gf_detail
        self.dbg['gfA'] = gf_a

        self.dbg['Igc'] = self.compute_brightness_guide(img)

        base_out = self.params.k1 * self.dbg['NStretch'] * (img - self.dbg['NMean']) \
                   + self.params.k2 * self.dbg['Igc']   # Eq. 6
        self.dbg['baseOut'] = base_out

        abs_a = np.abs(gf_a)
        max_a = float(abs_a.max())
        if max_a < 1e-9:
            max_a = 1.0
        norm_a = abs_a / max_a                       # 0..1
        mask = self.params.gLow + (self.params.gHigh - self.params.gLow) * norm_a
        detail_out = gf_detail * abs_a * mask         # Eq. 7
        self.dbg['detailOut'] = detail_out

        out32f = base_out + self.params.dde * detail_out   # Eq. 8

        clipped = np.clip(out32f, 0.0, 255.0)
        return np.round(clipped).astype(np.uint8)
